File: src/hooks/teammate-idle.ts
import { extractCwd, isConfigured, listPendingTasks, loadHookConfig, postMemoryEvent, readHookInput, writeHookOutput } from "./common.js";

function firstString(input: Record<string, unknown>, keys: string[]): string | undefined {
  for (const key of keys) {
    const value = input[key];
    if (typeof value === "string" && value.trim()) return value.trim();
  }
  return undefined;
}

export async function runTeammateIdleHook(): Promise<void> {
  const input = readHookInput<Record<string, unknown>>();
  const cwd = extractCwd(input);
  const config = loadHookConfig(cwd);

  const teammateName =
    firstString(input, ["teammate_name", "teammateName", "agent_name"]) || "teammate";
  const teamName = firstString(input, ["team_name", "teamName"]) || "team";

  if (!isConfigured(config)) {
    writeHookOutput();
    return;
  }

  const pendingTasks = await listPendingTasks(config, 5);

  await postMemoryEvent(
    config,
    "Teammate idle",
    {
      teammate_name: teammateName,
      team_name: teamName,
      pending_tasks: pendingTasks.slice(0, 5),
      timestamp: new Date().toISOString(),
    },
    ["hook", "teammate", "idle"]
  ).catch(() => {});

  const shouldRedirect = process.env.CONTEXTSTREAM_TEAMMATE_IDLE_REDIRECT !== "false";
  if (shouldRedirect && pendingTasks.length > 0) {
    const firstTask = pendingTasks[0] as Record<string, unknown>;
    const taskTitle =
      (typeof firstTask.title === "string" && firstTask.title) ||
      (typeof firstTask.subject === "string" && firstTask.subject) ||
      (typeof firstTask.name === "string" && firstTask.name) ||
      "pending task";
    const message = `Pending ContextStream task available: ${taskTitle}. Continue and complete this task before idling.`;
    writeHookOutput({ additionalContext: message, blocked: true, reason: message });
    return;
  }

  writeHookOutput();
}

const isDirectRun =
  process.argv[1]?.includes("teammate-idle") || process.argv[2] === "teammate-idle";
if (isDirectRun) {
  runTeammateIdleHook().catch(() => process.exit(0));
}

File: crates/mcp-server/src/hook_handlers/teammate_idle.rs
//! TeammateIdle hook handler.
//!
//! Captures teammate idle events and can redirect teammates to pending tasks.

use anyhow::Result;
use serde_json::Value;

use super::common::{extract_cwd, list_pending_tasks, load_config, post_memory_event};
use super::{read_stdin_json, write_stdout_json, HookOutput};

/// Handle the TeammateIdle hook.
pub async fn handle() -> Result<()> {
    let input = read_stdin_json()?;
    let cwd = extract_cwd(&input);
    let config = load_config(&cwd);

    let teammate_name = first_str(&input, &["teammate_name", "teammateName", "agent_name"])
        .unwrap_or("teammate")
        .to_string();
    let team_name = first_str(&input, &["team_name", "teamName"]).unwrap_or("team");

    if !config.is_configured() {
        write_stdout_json(&HookOutput::empty())?;
        return Ok(());
    }

    let pending_tasks = list_pending_tasks(&config, 5).await;

    let summary = serde_json::json!({
        "teammate_name": teammate_name,
        "team_name": team_name,
        "pending_tasks": pending_tasks.iter().take(5).collect::<Vec<_>>(),
        "timestamp": chrono::Utc::now().to_rfc3339(),
    });

    post_memory_event(
        &config,
        "Teammate idle",
        summary,
        &["hook", "teammate", "idle"],
    )
    .await;

    let should_redirect = std::env::var("CONTEXTSTREAM_TEAMMATE_IDLE_REDIRECT")
        .map(|v| v != "false")
        .unwrap_or(true);

    if should_redirect {
        if let Some(task) = pending_tasks.first() {
            let task_title = task
                .get("title")
                .or_else(|| task.get("subject"))
                .or_else(|| task.get("name"))
                .and_then(|v| v.as_str())
                .unwrap_or("pending task");

            let message = format!(
                "Pending ContextStream task available: {}. Continue and complete this task before idling.",
                task_title
            );

            eprintln!("{}", message);
            std::process::exit(2);
        }
    }

    write_stdout_json(&HookOutput::empty())?;
    Ok(())
}

fn first_str<'a>(value: &'a Value, keys: &[&str]) -> Option<&'a str> {
    keys.iter()
        .find_map(|k| value.get(*k).and_then(|v| v.as_str()))
}
